### ISSUE TITLE
New optimizer tests are flaky

### ISSUE THREAD

**Issue by chrisburr** [DIRACGrid/DIRAC#6472]
The new optimizer tests added in https://github.com/DIRACGrid/DIRAC/pull/6129 seem to be flaky:

For example they failed in these two nightly tests of DIRACOS2:

* https://github.com/DIRACGrid/DIRACOS2/actions/runs/3353998996/jobs/5557386007
* https://github.com/DIRACGrid/DIRACOS2/actions/runs/3358441562/jobs/5565461003

**Comment by chrisburr** [comment]
There seems to be at least two issues, one causes the test to hang until it times out after 5 minutes:

```python
2022-10-31 07:59:58 UTC WorkloadManagement/OptimizationMind NOTICE: Returning response unknown (0.00 secs) OK
2022-10-31 07:59:58 UTC WorkloadManagement/OptimizationMind NOTICE: Executing action unknown Connection/drop(<masked>)
2022-10-31 07:59:58 UTC WorkloadManagement/OptimizationMind WARN: Triggered conn_drop for ::ffff:172.23.0.5:9175->::ffff:172.23.0.5:39618
2022-10-31 07:59:58 UTC WorkloadManagement/OptimizationMind/ExecutorDispatcher INFO: Removing executor ::ffff:172.23.0.5:9175->::ffff:172.23.0.5:39618
2022-10-31 07:59:58 UTC WorkloadManagement/OptimizationMind NOTICE: Executing action ([::ffff:172.23.0.5]:39632)[hosts:server] Connection/new(<masked>)
2022-10-31 07:59:58 UTC WorkloadManagement/OptimizationMind NOTICE: Returning response ([::ffff:172.23.0.5]:39632)[hosts:server] (0.04 secs) OK
2022-10-31 07:59:58 UTC WorkloadManagement/OptimizationMind NOTICE: Executing action ([::ffff:172.23.0.5]:39632)[hosts:server] Connection/connected(<masked>)
2022-10-31 07:59:58 UTC WorkloadManagement/OptimizationMind ERROR: Exception while executing prepareToSend: 13
Traceback (most recent call last):
  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/Core/Base/ExecutorMindHandler.py", line 97, in __sendTask
    result = self.exec_prepareToSend(taskId, taskObj, eId)
  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/WorkloadManagementSystem/Service/OptimizationMindHandler.py", line 183, in exec_prepareToSend
    return jobState.recheckValidity()
  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/WorkloadManagementSystem/Client/JobState/CachedJobState.py", line 43, in recheckValidity
    result = self.__jobState.getAttributes(["Status", "MinorStatus", "LastUpdateTime"])
  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/WorkloadManagementSystem/Client/JobState/JobState.py", line 223, in getAttributes
    return JobState.__db.jobDB.getJobAttributes(self.__jid, nameList)
  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/WorkloadManagementSystem/DB/JobDB.py", line 282, in getJobAttributes
    return S_OK(result["Value"][jobID])
KeyError: 13
2022-10-31 07:59:58 UTC WorkloadManagement/OptimizationMind/ExecutorDispatcher ERROR: Failed to cbSendTask {'OK': False, 'Errno': 0, 'Message': 'Cannot presend task', 'CallStack': ['  File "/home/dirac/ServerInstallDIR/diracos/lib/python3.9/thre
ading.py", line 937, in _bootstrap\n    self._bootstrap_inner()\n', '  File "/home/dirac/ServerInstallDIR/diracos/lib/python3.9/threading.py", line 980, in _bootstrap_inner\n    self.run()\n', '  File "/home/dirac/ServerInstallDIR/diracos/lib/py
thon3.9/threading.py", line 917, in run\n    self._target(*self._args, **self._kwargs)\n', '  File "/home/dirac/ServerInstallDIR/diracos/lib/python3.9/concurrent/futures/thread.py", line 83, in _worker\n    work_item.run()\n', '  File "/home/dir
ac/ServerInstallDIR/diracos/lib/python3.9/concurrent/futures/thread.py", line 58, in run\n    result = self.fn(*self.args, **self.kwargs)\n', '  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/Core/DISET/private/Service.py", line
 344, in _processInThread\n    result = self._processProposal(trid, proposalTuple, handlerObj)\n', '  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/Core/DISET/private/Service.py", line 534, in _processProposal\n    result = sel
f._mbConnect(trid, handlerObj)\n', '  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/Core/DISET/private/Service.py", line 547, in _mbConnect\n    return handlerObj._rh_executeConnectionCallback("connected")\n', '  File "/home/di
rac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/Core/DISET/RequestHandler.py", line 404, in _rh_executeConnectionCallback\n    uReturnValue = oMethod(self.__trid, *args)\n', '  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/Co
re/Base/ExecutorMindHandler.py", line 149, in conn_connected\n    self.__eDispatch.addExecutor(trid, kwargs["executorTypes"])\n', '  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/Core/Utilities/ExecutorDispatcher.py", line 377,
 in addExecutor\n    self.__fillExecutors(eType)\n', '  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/Core/Utilities/ExecutorDispatcher.py", line 735, in __fillExecutors\n    result = self.__sendTaskToExecutor(eId, eType)\n', '
  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/Core/Utilities/ExecutorDispatcher.py", line 771, in __sendTaskToExecutor\n    self.__msgTaskToExecutor(taskId, eId, eType)\n', '  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/D
IRAC/src/DIRAC/Core/Utilities/ExecutorDispatcher.py", line 782, in __msgTaskToExecutor\n    result = self.__cbHolder.cbSendTask(taskId, self.__tasks[taskId].taskObj, eId, eType)\n', '  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DI
RAC/Core/Base/ExecutorMindHandler.py", line 42, in cbSendTask\n    return self.__sendTaskCB(taskId, taskObj, eId, eType)\n', '  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/Core/Base/ExecutorMindHandler.py", line 102, in __sen
dTask\n    return S_ERROR("Cannot presend task")\n']}
2022-10-31 07:59:58 UTC WorkloadManagement/OptimizationMind/ExecutorDispatcher ERROR: Exception while sending task to executor
Traceback (most recent call last):
  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/Core/Utilities/ExecutorDispatcher.py", line 771, in __sendTaskToExecutor
    self.__msgTaskToExecutor(taskId, eId, eType)
  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/Core/Utilities/ExecutorDispatcher.py", line 789, in __msgTaskToExecutor
    raise RuntimeError(result)
RuntimeError: {'OK': False, 'Errno': 0, 'Message': 'Cannot presend task', 'CallStack': ['  File "/home/dirac/ServerInstallDIR/diracos/lib/python3.9/threading.py", line 937, in _bootstrap\n    self._bootstrap_inner()\n', '  File "/home/dirac/Serv
erInstallDIR/diracos/lib/python3.9/threading.py", line 980, in _bootstrap_inner\n    self.run()\n', '  File "/home/dirac/ServerInstallDIR/diracos/lib/python3.9/threading.py", line 917, in run\n    self._target(*self._args, **self._kwargs)\n', '
 File "/home/dirac/ServerInstallDIR/diracos/lib/python3.9/concurrent/futures/thread.py", line 83, in _worker\n    work_item.run()\n', '  File "/home/dirac/ServerInstallDIR/diracos/lib/python3.9/concurrent/futures/thread.py", line 58, in run\n
 result = self.fn(*self.args, **self.kwargs)\n', '  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/Core/DISET/private/Service.py", line 344, in _processInThread\n    result = self._processProposal(trid, proposalTuple, handlerObj
)\n', '  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/Core/DISET/private/Service.py", line 534, in _processProposal\n    result = self._mbConnect(trid, handlerObj)\n', '  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/s
rc/DIRAC/Core/DISET/private/Service.py", line 547, in _mbConnect\n    return handlerObj._rh_executeConnectionCallback("connected")\n', '  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/Core/DISET/RequestHandler.py", line 404, in
 _rh_executeConnectionCallback\n    uReturnValue = oMethod(self.__trid, *args)\n', '  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/Core/Base/ExecutorMindHandler.py", line 149, in conn_connected\n    self.__eDispatch.addExecuto
r(trid, kwargs["executorTypes"])\n', '  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/Core/Utilities/ExecutorDispatcher.py", line 377, in addExecutor\n    self.__fillExecutors(eType)\n', '  File "/home/dirac/LocalRepo/ALTERNATI
VE_MODULES/DIRAC/src/DIRAC/Core/Utilities/ExecutorDispatcher.py", line 735, in __fillExecutors\n    result = self.__sendTaskToExecutor(eId, eType)\n', '  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/Core/Utilities/ExecutorDisp
atcher.py", line 771, in __sendTaskToExecutor\n    self.__msgTaskToExecutor(taskId, eId, eType)\n', '  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/Core/Utilities/ExecutorDispatcher.py", line 782, in __msgTaskToExecutor\n    r
esult = self.__cbHolder.cbSendTask(taskId, self.__tasks[taskId].taskObj, eId, eType)\n', '  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/Core/Base/ExecutorMindHandler.py", line 42, in cbSendTask\n    return self.__sendTaskCB(t
askId, taskObj, eId, eType)\n', '  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/Core/Base/ExecutorMindHandler.py", line 102, in __sendTask\n    return S_ERROR("Cannot presend task")\n']}
2022-10-31 07:59:58 UTC WorkloadManagement/OptimizationMind/ExecutorDispatcher ERROR: Could not send task to executor Exception while sending task to executor
2022-10-31 07:59:58 UTC WorkloadManagement/OptimizationMind INFO: Getting jobs for WorkloadManagement/JobScheduling
2022-10-31 07:59:58 UTC WorkloadManagement/OptimizationMind INFO: Added 0/0 jobs for Checking state
2022-10-31 07:59:59 UTC WorkloadManagement/OptimizationMind NOTICE: Executing action ([::ffff:172.23.0.5]:39632)[hosts:server] Message/msg_ExecutorError(<masked>)
2022-10-31 07:59:59 UTC WorkloadManagement/OptimizationMind INFO: Disconnecting executor by error: Error processing task 14: 14
2022-10-31 07:59:59 UTC WorkloadManagement/OptimizationMind/ExecutorDispatcher INFO: Removing executor ::ffff:172.23.0.5:9175->::ffff:172.23.0.5:39632
2022-10-31 07:59:59 UTC WorkloadManagement/OptimizationMind NOTICE: Returning response unknown (0.00 secs) OK
2022-10-31 07:59:59 UTC WorkloadManagement/OptimizationMind NOTICE: Executing action unknown Connection/drop(<masked>)
2022-10-31 07:59:59 UTC WorkloadManagement/OptimizationMind WARN: Triggered conn_drop for ::ffff:172.23.0.5:9175->::ffff:172.23.0.5:39632
```

Looking in the `JobScheduling` log I see errors like

```python
2022-10-31 07:49:57 UTC WorkloadManagement/JobScheduling ERROR: Error while processing task 12
Traceback (most recent call last):
  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/Core/Base/ExecutorReactor.py", line 178, in __moduleProcess
    result = modInstance._ex_processTask(taskId, taskStub)
  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/Core/Base/ExecutorModule.py", line 139, in _ex_processTask
    result = self.processTask(taskId, taskObj)
  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/WorkloadManagementSystem/Executor/Base/OptimizerExecutor.py", line 50, in processTask
    optResult = self.optimizeJob(jid, jobState)
  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/WorkloadManagementSystem/Executor/JobScheduling.py", line 67, in optimizeJob
    return self.__holdJob(jobState, "On Hold: after rescheduling %s" % reschedules, delay)
  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/WorkloadManagementSystem/Executor/JobScheduling.py", line 310, in __holdJob
    return jobState.setStatus(appStatus=holdMsg, source=self.ex_optimizerName())
TypeError: setStatus() missing 1 required positional argument: 'majorStatus'
2022-10-31 07:49:57 UTC WorkloadManagement/JobScheduling WARN: Error while receiving message from (172.23.0.5:9175) : Peer closed connection
```

```python
Traceback (most recent call last):
  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/Core/Base/ExecutorReactor.py", line 178, in __moduleProcess
    result = modInstance._ex_processTask(taskId, taskStub)
  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/Core/Base/ExecutorModule.py", line 139, in _ex_processTask
    result = self.processTask(taskId, taskObj)
  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/WorkloadManagementSystem/Executor/Base/OptimizerExecutor.py", line 50, in processTask
    optResult = self.optimizeJob(jid, jobState)
  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/WorkloadManagementSystem/Executor/JobScheduling.py", line 54, in optimizeJob
    result = jobState.getAttributes(["RescheduleCounter", "RescheduleTime", "ApplicationStatus"])
  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/WorkloadManagementSystem/Client/JobState/CachedJobState.py", line 334, in getAttributes
    return self.__cacheDict("att", self.__jobState.getAttributes, nameList)
  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/WorkloadManagementSystem/Client/JobState/CachedJobState.py", line 223, in __cacheDict
    result = functor(keyList)
  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/WorkloadManagementSystem/Client/JobState/JobState.py", line 223, in getAttributes
    return JobState.__db.jobDB.getJobAttributes(self.__jid, nameList)
  File "/home/dirac/LocalRepo/ALTERNATIVE_MODULES/DIRAC/src/DIRAC/WorkloadManagementSystem/DB/JobDB.py", line 282, in getJobAttributes
    return S_OK(result["Value"][jobID])
KeyError: 71
```

The test also fails fails in a different way to the CI if it's ran in a loop:

```python
            # Request job until ours is picked up or request returns S_ERROR
            while True:
                time.sleep(1)
                res = matcherClient.requestJob(resourceDescription)
                print(f"Matcher result: {res}")
                if not res["OK"] or res["Value"]["JobID"] == jobID:
                    break

            # Assert that our job has been selected by the matcher
>           assert res["OK"], res["Message"]
E           AssertionError: No match found ( 1510 : )
E           assert False

LocalRepo/TestCode/DIRAC/tests/Integration/WorkloadManagementSystem/Test_Client_WMS.py:219: AssertionError
```

**Comment by fstagni** [comment]
One fix is in https://github.com/DIRACGrid/DIRAC/pull/6473 but I am not sure it will be enough. I will leave this one open. 

For my info: tests in DIRACOS2 are running for which release? What you pointed above was for the pre-release.

**Comment by chrisburr** [comment]
They run for everything that I'm aware of people using:
* `rel-v7r2` (Py2 server)
* `rel-v7r3` (Py2 server)
* `rel-v7r3` (Py3 server)
* `integration`
* (as of this morning) `rel-v8r0`.

**Comment by chrisburr** [comment]
And another while making the release:

```
E           AssertionError: assert 'JobScheduling' == 'Pilot Agent Submission'
E             - Pilot Agent Submission
E             + JobScheduling
```

**Comment by chaen** [comment]
That one shows here too in DIRACOS2
https://github.com/DIRACGrid/DIRACOS2/actions/runs/3395480486/jobs/5656488217

**Comment by chrisburr** [comment]
Looks like they still are after #6574 https://gitlab.cern.ch/cburr/LHCbDIRAC/-/jobs/26018407/

**Comment by fstagni** [comment]
This means that the Optimizers did not run for 5 minutes. I will look what can be done.

**Comment by chrisburr** [comment]
I've seen this a couple more times...
```
             for jobID in jobIDList:
>               assert result["Value"][jobID]["Status"] == JobStatus.WAITING
E               AssertionError: assert 'Received' == 'Waiting'
E                 - Waiting
E                 + Received

LocalRepo/TestCode/DIRAC/tests/Integration/WorkloadManagementSystem/Test_Client_WMS.py:278: AssertionError
```

**Comment by chaen** [comment]
I too see this more often than not, and it is quite annoying. 

@fstagni did you make any progress on fixing that ?

**Comment by chaen** [comment]
unfortunately the problem is still there (https://gitlab.cern.ch/chaen/LHCbDIRAC/-/jobs/27304404)

```python
=================================== FAILURES ===================================
_________________________ test_submitJob_parametricJob _________________________

    def test_submitJob_parametricJob():
        """This test will submit a parametric job which should generate 3 actual jobs"""
    
        # create the job
        job = parametricJob()
    
        res = dirac.submitJob(job)
        assert res["OK"], res["Message"]
        jobIDList = res["Value"]
    
        try:
            assert len(jobIDList) == 3
    
            res = jobMonitoringClient.getJobsParameters(jobIDList, ["JobName"])
            assert res["OK"], res["Message"]
            jobNames = [res["Value"][jobID]["JobName"] for jobID in res["Value"]]
            assert set(jobNames) == {f"parametric_helloWorld_{nJob}" for nJob in range(3)}
    
            # Wait for the optimizers to run
            startingTime = time.time()
            while time.time() - startingTime < 7 * 60:
                result = jobMonitoringClient.getJobsStates(jobIDList)
                assert result["OK"], result["Message"]
                jobsAreNoLongerInChecking = True
                for jobID in jobIDList:
                    if result["Value"][jobID]["Status"] == JobStatus.CHECKING:
                        jobsAreNoLongerInChecking = False
                if jobsAreNoLongerInChecking:
                    break
    
                time.sleep(5)
            print(f"Lookup time: {time.time() - startingTime}s")
    
            for jobID in jobIDList:
>               assert result["Value"][jobID]["Status"] == JobStatus.WAITING
E               AssertionError: assert 'Received' == 'Waiting'
E                 - Waiting
E                 + Received

LocalRepo/TestCode/DIRAC/tests/Integration/WorkloadManagementSystem/Test_Client_WMS.py:276: AssertionError
```

**Comment by chaen** [comment]
the exact same in vanilla DIRAC again https://github.com/DIRACGrid/DIRAC/actions/runs/4104207760/jobs/7079383216

**Comment by chaen** [comment]
It keeps happening again and again. Could the priority of this be bumped ?
